cmd/kubeadm/app/apis/kubeadm/v1alpha3/doc.go

```diff
@@ -19,127 +19,255 @@ limitations under the License.
+//
+//     apiVersion: kubeadm.k8s.io/v1alpha3
+//     kind: InitConfiguration
+//     bootstrapTokens:
+//         ...
+//     nodeRegistration:
+//         ...
+//     apiEndpoint:
+//         ...
+//
+// InitConfiguration (and as well ClusterConfiguration afterwards) are originated from MasterConfiguration type
+// in v1alpha2 kubeadm config version.
+//
+// The InitConfiguration type should be used to configure runtime settings, that in case of kubeadm init
+// are the configuration of the bootstrap token and all the setting which are specific to the node where kubeadm
+// is executed, including:
+//
+// NodeRegistration, that holds fields that relate to registering the new node to the cluster;
+// use it to customize the node name, the CRI socket to use or any other settings that should apply to this
+// node only (e.g. the node ip).
+//
+// APIEndpoint, that represents the endpoint of the instance of the API server to be deployed on this node;
+// use it e.g. to customize the API server advertise address.
+//
+//     apiVersion: kubeadm.k8s.io/v1alpha3
+//     kind: ClusterConfiguration
+//     networking:
+//         ...
+//     etcd:
+//         ...
+//     apiServerExtraArgs:
+//         ...
+//     APIServerExtraVolumes:
+//         ...
+//     ...
+//
+// The ClusterConfiguration type should be used to configure cluster-wide settings,
+// including settings for:
+//
+// Networking, that holds configuration for the networking topology of the cluster; use it e.g. to customize
+// node subnet or services subnet.
+//
+// Etcd configurations; use it e.g. to customize the local etcd or to configure the API server
+// for using an external etcd cluster.
+//
+// kube-apiserver, kube-scheduler, kube-controller-manager configurations; use it to customize control-plane
+// components by adding customized setting or overriding kubeadm default settings.
+//
+//    apiVersion: kubeproxy.config.k8s.io/v1alpha1
+//    kind: KubeProxyConfiguration
+//       ...
+//
+// The KubeProxyConfiguration type should be used to change the configuration passed to kube-proxy instances deployed
+// in the cluster. If this object is not provided or provided only partially, kubeadm applies defaults.
+//
+// See https://kubernetes.io/docs/reference/command-line-tools-reference/kube-proxy/ or https://godoc.org/k8s.io/kube-proxy/config/v1alpha1#KubeProxyConfiguration
+// for kube proxy official documentation.
+//
+//    apiVersion: kubelet.config.k8s.io/v1beta1
+//    kind: KubeletConfiguration
+//       ...
+//
+// The KubeletConfiguration type should be used to change the configurations that will be passed to all kubelet instances
+// deployed in the cluster. If this object is not provided or provided only partially, kubeadm applies defaults.
+//
+// See https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/ or https://godoc.org/k8s.io/kubelet/config/v1beta1#KubeletConfiguration
+// for kube proxy official documentation.
+//
```

Comment:
i think we should still include the previous text about what command supports what types?
```
For `kubeadm init` you can include the following types:
//   InitConfiguration, ClusterConfiguration, KubeProxyConfiguration, KubeletConfiguration
//
// For `kubeadm join` you can include the following types:
//   JoinConfiguration, KubeProxyConfiguration, KubeletConfiguration
```

this was a confusing part for users.

Reply:
This is done at the beginning of chapters for kubeadm init and kubeadm join.
Please note that kubeadm join considers only JoinConfiguration (kubeadm reads from the cluster KubeProxyConfiguration and KubeletConfiguration, but the user doesn't have to provide them via config file)

Reply:
i was wondering about this.
so in a case someone wants to provide a custom KubeletConfiguration for a worker node, can they still include KubeletConfiguration next to their JoinConfiguration separated with ---?

Reply:
No
KubeletConfiguration can be used only for cluster wide kubelet settings and it is processed only by kubeadm init.
If you want to customize kubelet configuration on a node basis you can use kubelet extra args into the joinconfiguration.noderegistration object.

Reply:
This PR focuses on documenting v1alpha3, so this is out of scope.
But if there is a plan for flag deprecation, what you are saying should be definitely included in v1beta1 redesign.

Reply:
> This PR focuses on documenting v1alpha3, so this is out of scope.
yes, this is outside of the context for this PR,

> But if there is a plan for flag deprecation, what you are saying should be definitely included in v1beta1 redesign.

100%